src/ORM/Related/DataObjectRelatedService.php

```diff
@@ -0,0 +1,437 @@
+                return null;
+            }
+            $joinTableName = $this->classToTableName[$class];
+        }
+        return $joinTableName;
+    }
+
+    /**
+     * @param array $classIDs
+     * @param DataObject $record
+     * @param string $class
+     */
+    private function addRelatedReverseHasOnes(array &$classIDs, DataObject $record, string $class): void
+    {
+        foreach (singleton($class)->hasOne() as $component => $componentClass) {
+            if (!($record instanceof $componentClass)) {
+                continue;
+            }
+            $results = $this->fetchReverseHasOneResults($record, $class, $component);
+            $this->addResultsToClassIDs($classIDs, $results, $class);
+        }
+    }
+
+    /**
+     * Query the database to retrieve has-one results
+     *
+     * @param DataObject $record - The DataObject whose usage data is being retrieved, usually a File
+     * @param string $class - example: My\App\SomePageType
+     * @param string $component - example: 'SomeFiles' - My\App\SomePageType::SomeFiles()
+     * @return Query|null
+     */
+    private function fetchReverseHasOneResults(DataObject $record, string $class, string $component): ?Query
+    {
+        // Ensure table exists, this is required for TestOnly SapphireTest classes
+        if (!isset($this->classToTableName[$class])) {
+            return null;
+        }
+        $componentIDField = "{$component}ID";
+
+        // Only get database fields from the current class model, not parent class model
+        $dbFields = $this->dataObjectSchema->databaseFields($class, false);
+        if (!isset($dbFields[$componentIDField])) {
+            return null;
+        }
+        $tableName = $this->dataObjectSchema->tableName($class);
+
+        // Example SQL: SELECT "ID" FROM "MyPage" WHERE "MyFileID" = 123;
+        return SQLSelect::create(
+            '"ID"',
+            sprintf('"%s"', $tableName),
+            sprintf('"%s" = %s', $componentIDField, $record->ID)
+        )->execute();
+    }
+
+    /**
+     * @param array $classIDs
+     * @param DataObject $record
+     * @param string $class
+     * @param array $throughClasses
+     */
+    private function addRelatedReverseManyManys(
+        array &$classIDs,
+        DataObject $record,
+        string $class,
+        array &$throughClasses
+    ): void {
+        foreach (singleton($class)->manyMany() as $component => $componentClass) {
+            $componentClass = $this->updateComponentClass($componentClass, $throughClasses);
+            if (!($record instanceof $componentClass)) {
+                continue;
+            }
+            $results = $this->fetchManyManyResults($record, $class, $component, true);
+            $this->addResultsToClassIDs($classIDs, $results, $class);
+        }
+    }
+
+    /**
+     * Update the $classIDs array with the relationship IDs from database $results
+     *
+     * @param array $classIDs
```

Comment:
```suggestion
     * Update the `$classIDs` array with the relationship IDs from database `$results`.
     *
     * @param int[] $classIDs
```

Reply:
$classIDs isn't an int[], it's an assoc_array where $key = string (ClassName) and $value = int (ID)